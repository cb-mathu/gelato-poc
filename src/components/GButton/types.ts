import { VueComponentProps } from '../../types';
import { ButtonHTMLAttributes, ExtractPropTypes, VNodeProps } from 'vue';
import { BUTTON_PROPS } from './constants';

export type ButtonType = ButtonHTMLAttributes['type'];

export enum ButtonVariant {
  Primary = 'primary',
  Outlined = 'outlined',
  Link = 'link',
}

export type ButtonProps = ExtractPropTypes<typeof BUTTON_PROPS>;
export type NativeButtonProps = Omit<ButtonHTMLAttributes, keyof ButtonProps>;
export type MergedButtonProps = Partial<ButtonProps & VueComponentProps & NativeButtonProps>;
