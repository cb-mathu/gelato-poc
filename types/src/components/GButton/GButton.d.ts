import { MergedButtonProps } from './types';
declare const _default: new () => {
    $props: MergedButtonProps;
};
export default _default;
