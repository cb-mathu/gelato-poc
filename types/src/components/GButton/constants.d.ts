import { PropType } from 'vue';
import { ButtonVariant } from './types';
export declare const VARIANTS: ButtonVariant[];
export declare const BUTTON_PROPS: {
    as: {
        type: StringConstructor;
        default: string;
    };
    type: PropType<"submit" | "reset" | "button" | undefined>;
    variant: {
        type: PropType<ButtonVariant>;
        validator: (val: ButtonVariant) => boolean;
    };
    fluid: {
        type: BooleanConstructor;
        default: boolean;
    };
    loading: {
        type: BooleanConstructor;
        default: boolean;
    };
    loaderId: {
        type: StringConstructor;
        required: boolean;
    };
    fill: {
        type: StringConstructor;
        required: boolean;
    };
    id: {
        type: StringConstructor;
        required: boolean;
    };
    class: {
        type: StringConstructor;
        required: boolean;
        default: string;
    };
};
