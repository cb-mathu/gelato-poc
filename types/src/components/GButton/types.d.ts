import { VueComponentProps } from '../../types';
import { ButtonHTMLAttributes, ExtractPropTypes } from 'vue';
import { BUTTON_PROPS } from './constants';
export declare type ButtonType = ButtonHTMLAttributes['type'];
export declare enum ButtonVariant {
    Primary = "primary",
    Outlined = "outlined",
    Link = "link"
}
export declare type ButtonProps = ExtractPropTypes<typeof BUTTON_PROPS>;
export declare type NativeButtonProps = Omit<ButtonHTMLAttributes, keyof ButtonProps>;
export declare type MergedButtonProps = Partial<ButtonProps & VueComponentProps & NativeButtonProps>;
