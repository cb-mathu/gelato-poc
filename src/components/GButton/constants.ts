import { PropType } from 'vue';
import { DEFAULT_PROPS } from '../../utils/props';
import { ButtonVariant, ButtonType } from './types';

export const VARIANTS = [ButtonVariant.Primary, ButtonVariant.Outlined, ButtonVariant.Link];

export const BUTTON_PROPS = {
  ...DEFAULT_PROPS,
  as: {
    type: String,
    default: 'button',
  },
  type: String as PropType<ButtonType>,
  variant: {
    type: String as PropType<ButtonVariant>,
    validator: (val: ButtonVariant) => VARIANTS.includes(val),
  },
  fluid: {
    type: Boolean,
    default: false,
  },
  loading: {
    type: Boolean,
    default: false,
  },
  loaderId: {
    type: String,
    required: false,
  },
  fill: {
    type: String,
    required: false,
  },
};
