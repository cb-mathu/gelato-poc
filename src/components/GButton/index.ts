import GButton from './GButton';
import { ButtonVariant } from './types';

export { GButton, ButtonVariant };
