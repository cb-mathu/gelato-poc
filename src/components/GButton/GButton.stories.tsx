import GButton from './GButton';
import { VARIANTS } from './constants';
import { MergedButtonProps, ButtonVariant, ButtonProps } from './types';

export default {
  title: 'Component/GButton',
  component: GButton,
  argTypes: {
    fluid: {
      description: 'Make button takes full width available',
      control: {
        type: 'boolean',
      },
    },
    loading: {
      description: 'Show progress inside button',
      control: {
        type: 'boolean',
      },
    },
    variant: {
      description: 'Button variants',
      options: VARIANTS,
      control: {
        type: 'select',
      },
    },
  },
};

const Template = (args: ButtonProps) => ({
  setup() {
    return () => <GButton {...args}>Click me!</GButton>;
  },
});

export const Story = Template.bind({});
export const DefaultButton = (props?: MergedButtonProps) => <GButton {...(props as any)}>Click me!</GButton>;
export const LoadingButton = () => (
  <GButton variant={ButtonVariant.Primary} loading>
    Click me!
  </GButton>
);
export const PrimaryButton = () => <GButton variant={ButtonVariant.Primary}>Click me!</GButton>;
export const OutlinedButton = () => <GButton variant={ButtonVariant.Outlined}>Click me!</GButton>;
export const LinkButton = (props?: MergedButtonProps) => (
  <GButton variant={ButtonVariant.Link} onClick={props?.onClick}>
    Click me!
  </GButton>
);
