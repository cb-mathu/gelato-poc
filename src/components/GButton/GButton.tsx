import { defineComponent, SetupContext } from 'vue';
import { BUTTON_PROPS } from './constants';
import { ButtonVariant, MergedButtonProps, ButtonProps } from './types';
import { Spinner } from '../icons';

const GButton = defineComponent({
  name: 'GButton',
  props: BUTTON_PROPS,

  setup(props: ButtonProps, { slots, attrs }: SetupContext) {
    return () => {
      const classList = {
        [`g-button`]: true,
        [`g-button--primary`]: props.variant === ButtonVariant.Primary,
        [`g-button--link`]: props.variant === ButtonVariant.Link,
        [`g-button--fluid`]: props.fluid,
        [`g-button--loading`]: props.loading,
      };

      const attributes = {
        ...attrs,
        ...props,
      };

      return (
        <button {...attributes} class={classList}>
          {attributes.loading && <Spinner class="g-button__spinner" id={props?.loaderId} fill={props?.fill} />}
          <span class="g-button__label">{slots.default?.()}</span>
        </button>
      );
    };
  },
});

export default GButton as {
  new (): {
    $props: MergedButtonProps;
  };
};
