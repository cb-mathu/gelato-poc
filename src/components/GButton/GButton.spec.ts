import { render, screen, fireEvent } from '@testing-library/vue';
import * as stories from './GButton.stories';

const { PrimaryButton, OutlinedButton, LinkButton, DefaultButton } = stories;

describe('GButton : ', () => {
  test('renders primary button with default args', () => {
    render(PrimaryButton());
    const buttonElement = screen.getByText('Click me!');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders outlined button with default args', () => {
    render(OutlinedButton());
    const buttonElement = screen.getByText('Click me!');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders link button with default args', () => {
    render(LinkButton());
    const buttonElement = screen.getByText('Click me!');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders default button as anchor tag', () => {
    render(DefaultButton({ as: 'a' }));
    const buttonElement = screen.getByText('Click me!');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders default button as fluid', () => {
    render(DefaultButton({ fluid: true }));
    const buttonElement = screen.getByText('Click me!');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders default button with native attributes passed', () => {
    render(DefaultButton({ id: 'my-button', class: 'my-button', disabled: true }));
    const buttonElement = screen.getByText('Click me!');
    expect(buttonElement).toMatchSnapshot();
  });

  test('calls onClick prop when clicked', () => {
    const handleClick = jest.fn();
    render(DefaultButton({ onClick: handleClick }));
    fireEvent.click(screen.getByText('Click me!'));
    expect(handleClick).toHaveBeenCalled();
  });
});
