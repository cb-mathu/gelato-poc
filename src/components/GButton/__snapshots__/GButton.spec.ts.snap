// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GButton :  renders default button as anchor tag 1`] = `
<span
  class="g-button__label"
>
  
  Click me!
  
</span>
`;

exports[`GButton :  renders default button as fluid 1`] = `
<span
  class="g-button__label"
>
  
  Click me!
  
</span>
`;

exports[`GButton :  renders default button with native attributes passed 1`] = `
<span
  class="g-button__label"
>
  
  Click me!
  
</span>
`;

exports[`GButton :  renders link button with default args 1`] = `
<span
  class="g-button__label"
>
  
  Click me!
  
</span>
`;

exports[`GButton :  renders outlined button with default args 1`] = `
<span
  class="g-button__label"
>
  
  Click me!
  
</span>
`;

exports[`GButton :  renders primary button with default args 1`] = `
<span
  class="g-button__label"
>
  
  Click me!
  
</span>
`;
